File: DSA-Sheets/Striver-SDE-Sheet/01-Arr-I/Sort-colors/dutch-national-flag-algo.cpp
// three pointers => 0th pointer 1th pointer 2th pointer
#include <bits/stdc++.h>
using namespace std;

class Solution
{
public:
    void sortColors(vector<int> &nums)
    {
        // dutch national flag algo => single iteration => O(n)
        int low = 0;
        int mid = 0;
        int high = nums.size() - 1;

        while (mid <= high)
        {
            if (nums[mid] == 0)
                swap(nums[mid++], nums[low++]);
            else if (nums[mid] == 1)
                mid++;
            else
                swap(nums[mid], nums[high--]);
        }
    }
};

File: DSA-Sheets/Striver-SDE-Sheet/01-Arr-I/Sort-colors/dutch-national-flag-algo.py
class Solution:
    def sortColors(self, nums: List[int]) -> None:
        # dutch national flag algorithm 
        low, mid, high = 0, 0, len(nums)-1

        while mid <= high :
            if nums[mid] == 0 :
                nums[low], nums[mid] = nums[mid], nums[low]
                low += 1
            elif nums[mid] == 2:
                nums[mid], nums[high] = nums[high], nums[mid]
                high -= 1
                mid -= 1
            mid += 1

